File: packages/mcp/src/parser/extractors/items.ts
import type { Item, ItemForm, RawClass } from '../types.js';
import { getNumber, getString } from '../util.js';

const FORM_MAP: Record<string, ItemForm> = {
  RF_SOLID: 'solid',
  RF_LIQUID: 'liquid',
  RF_GAS: 'gas',
  RF_INVALID: 'invalid',
};

/** Stack-size enum → count, used when `mCachedStackSize` is absent. */
const STACK_SIZE_MAP: Record<string, number> = {
  SS_ONE: 1,
  SS_SMALL: 50,
  SS_MEDIUM: 100,
  SS_BIG: 200,
  SS_HUGE: 500,
  SS_FLUID: 0,
};

export function mapForm(raw: string): ItemForm {
  return FORM_MAP[raw] ?? 'invalid';
}

function resolveStackSize(raw: RawClass): number {
  // `mCachedStackSize` is the resolved integer when present; prefer it.
  const cached = getNumber(raw, 'mCachedStackSize', Number.NaN);
  if (Number.isFinite(cached) && cached > 0) {
    return cached;
  }
  const enumValue = getString(raw, 'mStackSize');
  return STACK_SIZE_MAP[enumValue] ?? 0;
}

/**
 * Builds an `Item` from a raw item/resource class. `isResource` distinguishes
 * raw resources (leaf nodes) from manufactured items.
 */
export function itemFromRaw(raw: RawClass, isResource: boolean): Item {
  return {
    className: getString(raw, 'ClassName'),
    displayName: getString(raw, 'mDisplayName'),
    description: getString(raw, 'mDescription'),
    stackSize: resolveStackSize(raw),
    form: mapForm(getString(raw, 'mForm')),
    sinkPoints: getNumber(raw, 'mResourceSinkPoints', 0),
    isResource,
  };
}

File: packages/sf-game-data/extract/sf-game-data-parse/Items.cs
using System.Text.Json;

namespace SfGameData.Parse;

// Ported from src/parser/extractors/items.ts.
public static class Items
{
    private static readonly Dictionary<string, string> FormMap = new()
    {
        ["RF_SOLID"] = "solid",
        ["RF_LIQUID"] = "liquid",
        ["RF_GAS"] = "gas",
        ["RF_INVALID"] = "invalid",
    };

    // Stack-size enum → count, used when mCachedStackSize is absent.
    private static readonly Dictionary<string, double> StackSizeMap = new()
    {
        ["SS_ONE"] = 1,
        ["SS_SMALL"] = 50,
        ["SS_MEDIUM"] = 100,
        ["SS_BIG"] = 200,
        ["SS_HUGE"] = 500,
        ["SS_FLUID"] = 0,
    };

    public static string MapForm(string raw) => FormMap.GetValueOrDefault(raw, "invalid");

    private static double ResolveStackSize(JsonElement raw)
    {
        var cached = Util.GetNumber(raw, "mCachedStackSize", double.NaN);
        if (double.IsFinite(cached) && cached > 0)
        {
            return cached;
        }
        var enumValue = Util.GetString(raw, "mStackSize");
        return StackSizeMap.GetValueOrDefault(enumValue, 0);
    }

    public static Item ItemFromRaw(JsonElement raw, bool isResource) => new()
    {
        ClassName = Util.GetString(raw, "ClassName"),
        DisplayName = Util.GetString(raw, "mDisplayName"),
        Description = Util.GetString(raw, "mDescription"),
        StackSize = ResolveStackSize(raw),
        Form = MapForm(Util.GetString(raw, "mForm")),
        SinkPoints = Util.GetNumber(raw, "mResourceSinkPoints", 0),
        EnergyValue = Util.GetNumber(raw, "mEnergyValue", 0),
        IsResource = isResource,
    };
}
